`next-typesafe-url@^5` doesn't support Next.js 14

Next.js 15 updated `params` and `searchParams` to be `Promise`-s.

`next-typesafe-url@^5` assumes they are indeed `Promise`-s and doesn't handle non-`Promise` variant, leading to a `paramsPromise.then is not a function` error:

```
[1]  ⨯ node_modules/.pnpm/next-typesafe-url@5.1.2_next@14.1.0_@babel+core@7.23.6_@opentelemetry+api@1.7.0_babel-plugin-_wcl3gv6ubytpkg3rncjokoiojy/node_modules/next-typesafe-url/dist/app/hoc.mjs (20:38) @ then
[1]  ⨯ TypeError: paramsPromise.then is not a function
[1]     at stringify (<anonymous>)
```


Possible solutions:

* Require Next.js 15+ for `next-typesafe-url@^5` (dropping support for Next.js 14 and older)
* Adjust the code to handle non-`Promise` variants of `params` and `searchParams`

Almost working now, but I'm having issues with `/dashboard/[thingId]/edit/layout.tsx`:

```
.next/types/app/dashboard/[thingId]/edit/layout.ts:28:13 - error TS2344: Type 'OmitWithTag<Pick<NextAppPageProps, "params"> & { children: ReactNode; } & { routeParams: ReactNode; children: ReactNode; }, keyof LayoutProps, "default">' does not satisfy the constraint '{ [x: string]: never; }'.
  Property 'routeParams' is incompatible with index signature.
    Type 'ReactNode' is not assignable to type 'never'.
      Type 'undefined' is not assignable to type 'never'.

28 checkFields<Diff<LayoutProps, FirstArg<TEntry['default']>, 'default'>>()
               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

src/app/dashboard/[thingId]/edit/layout.tsx:60:42 - error TS2345: Argument of type '({ routeParams, children }: Props) => Element' is not assignable to parameter of type '(props: InferLayoutPropsType<Required<Pick<DynamicRoute, "routeParams">>, "routeParams" | "children">) => ReactNode | Promise<...>'.
  Types of parameters '__0' and 'props' are incompatible.
    Type 'InferLayoutPropsType<Required<Pick<DynamicRoute, "routeParams">>, "routeParams" | "children">' is not assignable to type 'Props'.
      Type 'InferLayoutPropsType<Required<Pick<DynamicRoute, "routeParams">>, "routeParams" | "children">' is not assignable to type '{ routeParams: Promise<{ thingId: string; }>; children: ReactNode; }'.
        Types of property 'routeParams' are incompatible.
          Type 'Promise<{ [x: string]: any; }> & ReactNode' is not assignable to type 'Promise<{ thingId: string; }>'.
            Type 'Promise<{ [x: string]: any; }> & string' is not assignable to type 'Promise<{ thingId: string; }>'.
              Types of property 'then' are incompatible.
                Type '<TResult1 = { [x: string]: any; }, TResult2 = never>(onfulfilled?: ((value: { [x: string]: any; }) => TResult1 | PromiseLike<TResult1>) | null | undefined, onrejected?: ((reason: any) => TResult2 | PromiseLike<...>) | null | undefined) => Promise<...>' is not assignable to type '<TResult1 = { thingId: string; }, TResult2 = never>(onfulfilled?: ((value: { thingId: string; }) => TResult1 | PromiseLike<TResult1>) | null | undefined, onrejected?: ((reason: any) => TResult2 | PromiseLike<...>) | null | undefined) => Promise<...>'.
                  Types of parameters 'onfulfilled' and 'onfulfilled' are incompatible.
                    Types of parameters 'value' and 'value' are incompatible.
                      Property 'thingId' is missing in type '{ [x: string]: any; }' but required in type '{ thingId: string; }'.

60 export default withLayoutParamValidation(Layout, LayoutRoute);
                                            ~~~~~~

  src/app/dashboard/[thingId]/edit/layout.tsx:17:5
    17     thingId: z.string(),
           ~~~~~~~~~~~~~~~~~~~~~~~~
    'thingId' is declared here.


Found 2 errors in 2 files.

Errors  Files
     1  .next/types/app/dashboard/[thingId]/edit/layout.ts:28
     1  src/app/dashboard/[thingId]/edit/layout.tsx:60
 ELIFECYCLE  Command failed with exit code 2.
```

Code itself:
```tsx
import { $path, DynamicLayout, InferLayoutPropsType } from "next-typesafe-url";
import { withLayoutParamValidation } from "next-typesafe-url/app/hoc";
import Link from "next/link";
import { Suspense } from "react";

import { z } from "zod";

import { Await } from "@/components/await";

import { EditTabs } from "./tabs";

////////////////////////////////////////////////////////////////////////////////////////////////////

const LayoutRoute = {
  routeParams: z.object({
    thingId: z.string(),
  }),
} satisfies DynamicLayout;
type LayoutType = typeof LayoutRoute;

////////////////////////////////////////////////////////////////////////////////////////////////////

type Props = InferLayoutPropsType<LayoutType>;

function Layout({ routeParams, children }: Props) {
  return (
    <>
      <Suspense>
        <Await promise={routeParams}>
          {(routeParams) => (
            <EditTabs thingId={routeParams.thingId} />
          )}
        </Await>
      </Suspense>

      {children}
    </>
  );
}

export default withLayoutParamValidation(Layout, LayoutRoute);
```

--- a/www/docs/src/content/docs/en/usage/search-route-params-app.md
+++ b/www/docs/src/content/docs/en/usage/search-route-params-app.md
@@ -80,6 +80,8 @@ The `InferLayoutPropsType` helper type is passed the type of your LayoutRoute as
 
 ---
 
+**Note:** It is required to pass the type of your LayoutRoute to `withLayoutParamValidation` due to type inference limitations.
+
 **Note:** If no error is thrown, your PageProps will always be the types you defined in your schema.
 
 **Note:** If your layout does not consume any route params, there is no need to use `withLayoutParamValidation`
@@ -109,7 +111,7 @@ async function Layout({ children, routeParams }: Props) {
   );
 }
 
-export default withLayoutParamValidation(Layout, LayoutRoute);
+export default withLayoutParamValidation<LayoutType>(Layout, LayoutRoute);
 ```
 
 #### Errors
@@ -193,7 +195,10 @@ function Layout({ children, routeParams, analytics }: Props) {
     </div>
   );
 }
-export default withLayoutParamValidation<"analytics">(Layout, LayoutRoute);
+export default withLayoutParamValidation<LayoutType, "analytics">(
+  Layout,
+  LayoutRoute
+);
 ```
 
 ### Intercepted Routes

